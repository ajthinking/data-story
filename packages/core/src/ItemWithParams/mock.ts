export const removePropertyData = {
  'name': 'remove_properties',
  'type': 'RepeatableParam',
  'row': [
    {
      'name': 'property',
      'type': 'StringableParam',
      'value': 'id'
    }
  ],
  'value': [
    {
      'property': 'foo-1'
    }
  ]
};
export const mockRepeatableData = {
  name: 'repeatableTest',
  type: 'RepeatableParam',
  row: [
    {
      name: 'value',
      type: 'StringableParam',
      value: 'id'
    },
    {
      ...removePropertyData
    }
  ],
  value: [
    {
      value: 'value-11',
      remove_properties: [
        {
          'property': 'property-11'
        }
      ]
    },
    {
      value: 'value-22',
      remove_properties: [
        {
          'property': 'property-22'
        }
      ]
    }
  ]
}

export const mockPortMapData = {
  'name': 'port_map',
  'type': 'RepeatableParam',
  'row': [
    {
      'name': 'value',
      'type': 'StringableParam',
      'value': 'id'
    },
    {
      'name': 'port',
      'type': 'PortSelectionParam',
      'value': '',
    }
  ],
  'value': [
    {
      'value': '22',
      'port': 'unfiltered'
    },
    {
      'value': 'id',
      'port': 'outputzdbj'
    }
  ]
};
