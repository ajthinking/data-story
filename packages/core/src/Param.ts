export type ParamValue = any

export type Param = {
  id: string
  name: string
  rows: number
  type: string
  value?: ParamValue
  placeholder?: string,
  selectOptions?: string[],
  inputSchemaFromPort?: string,
}

export const name: Param = {
  id: 'name',
  name: 'name',
  type: 'string',
  value: '',
  rows: 1,
}

export const label: Param = {
  id: 'label',
  name: 'label',
  type: 'string',
  value: '',
  rows: 1,  
}

export const DefaultParams = { name, label }