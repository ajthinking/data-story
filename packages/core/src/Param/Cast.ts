export type Cast = {
  type: string,
  label: string,
  selected?: boolean,
}