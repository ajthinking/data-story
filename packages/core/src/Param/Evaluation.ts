export type Evaluation = {
  type: string,
  label: string,
  shortLabel: string,
  selected?: boolean,
}