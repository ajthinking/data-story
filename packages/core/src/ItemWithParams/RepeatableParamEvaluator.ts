import { Param, RepeatableParam } from '../Param';
import { ItemValue } from '../types/ItemValue';
import { ParamsValueEvaluator } from '../types/ParamsValueEvaluator';

export class RepeatableParamEvaluator implements ParamsValueEvaluator<RepeatableParam<any>> {

  type = 'RepeatableParam' as const;

  constructor(private evaluator: ParamsValueEvaluator<any>) {
  }

  evaluate(itemValue: ItemValue, param: RepeatableParam<any>) {
    return Object.fromEntries(param.row.map((row: {name: string}) => {
      return [row.name, this.evaluator.evaluate(itemValue, row)] as const;
    }));
  }

  canEvaluate(param: Param): param is RepeatableParam<any> {
    return param.type === this.type;
  }
}
