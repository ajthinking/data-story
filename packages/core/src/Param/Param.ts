import { PortName } from '../types/Port'
import { Cast } from './Cast'
import { Evaluation } from './Evaluation'
import { numberCast } from './casts/numberCast'
import { stringCast } from './casts/stringCast'
import { jsExpressionEvaluation } from './evaluations/jsExpressionEvaluation'
import { jsFunctionEvaluation } from './evaluations/jsFunctionEvaluation'
import { jsonEvaluation } from './evaluations/jsonEvaluation'

export interface StringableInputValue extends Record<string, any> {
  rawValue: any,
  Evaluation?: string,
  Cast?: string,
}

export type StringableParam = {
  name: string,
  label: string,
  help: string,
  type: 'StringableParam',
  multiline: boolean,
  canInterpolate: boolean,
  interpolate?: boolean,
  interpolationsFromPort?: PortName[],
  casts?: Cast[],
  evaluations?: Evaluation[],
  input: StringableInputValue,
}

export type PortSelectionParam = {
  name: string,
  label: string,
  help: string,
  type: 'PortSelectionParam',
  allowCreate: boolean,
  input: string
}

/**
 * This type can represent ["a", "b", "c"] using 'a, b, c', which facilitates user input
 * */
export type StringListParam = {
  name: string,
  label: string,
  help: string,
  type: 'StringListParam',
  input: unknown
}

export type Param =
  StringableParam |
  PortSelectionParam |
  StringListParam;

export type ParamValue = Param['input']

export const str = ({
  name,
  label,
  help,
  multiline,
  canInterpolate,
  interpolate,
  evaluations,
  value,
}: {
  name: string,
  label?: string,
  help?: string,
  multiline?: boolean,
  canInterpolate?: boolean,
  interpolate?: boolean,
  evaluations?: Evaluation[],
  value?: string,
}): StringableParam => {
  return {
    name,
    type: 'StringableParam',
    label: label ?? name,
    help: help ?? '',
    multiline: multiline ?? false,
    canInterpolate: canInterpolate ?? true,
    interpolate: interpolate ?? true,
    evaluations: evaluations,
    casts: [
      stringCast,
    ],
    input: {
      rawValue: value ?? '',
      Cast: stringCast.type,
    },
  }
}

export const num = ({
  name,
  label,
  help,
  multiline,
  canInterpolate,
  interpolate,
  evaluations,
  value,
}: {
  name: string,
  label?: string,
  help?: string,
  multiline?: boolean,
  canInterpolate?: boolean,
  interpolate?: boolean,
  evaluations?: Evaluation[],
  value?: string | number,
}): Param => {
  return {
    name,
    type: 'StringableParam',
    label: label ?? name,
    help: help ?? '',
    multiline: multiline ?? false,
    canInterpolate: canInterpolate ?? true,
    interpolate: interpolate ?? true,
    evaluations: evaluations,
    casts: [
      numberCast,
    ],
    input: {
      rawValue: value ?? 0,
      Cast: numberCast.type,
    },
  }
}

export const json_ = ({
  name,
  label,
  help,
  multiline = true,
  canInterpolate,
  interpolate,
  evaluations,
  value,
}: {
  name: string,
  label?: string,
  help?: string,
  multiline?: boolean,
  canInterpolate?: boolean,
  interpolate?: boolean,
  evaluations?: Evaluation[],
  value?: string,
}): Param => {
  return {
    name,
    type: 'StringableParam',
    label: label ?? name,
    help: help ?? '',
    multiline: multiline ?? false,
    canInterpolate: canInterpolate ?? true,
    interpolate: interpolate ?? true,
    evaluations: evaluations ?? [
      jsonEvaluation,
      jsFunctionEvaluation,
      jsExpressionEvaluation,
    ],
    casts: [
      numberCast,
      stringCast,
    ],
    input: {
      rawValue: value ?? 0,
      Evaluation: jsonEvaluation.type,
    },
  }
}

export const jsFn = ({
  name,
  label,
  help,
  multiline = true,
  canInterpolate,
  interpolate,
  evaluations,
  value,
}: {
  name: string,
  label?: string,
  help?: string,
  multiline?: boolean,
  canInterpolate?: boolean,
  interpolate?: boolean,
  evaluations?: Evaluation[],
  value?: string,
}): Param => {
  return {
    name,
    type: 'StringableParam',
    label: label ?? name,
    help: help ?? '',
    multiline: multiline ?? false,
    canInterpolate: canInterpolate ?? true,
    interpolate: interpolate ?? true,
    evaluations: evaluations ?? [
      jsonEvaluation,
      jsFunctionEvaluation,
      jsExpressionEvaluation,
    ],
    casts: [
      numberCast,
      stringCast,
    ],
    input: {
      rawValue: value ?? 0,
      Evaluation: jsFunctionEvaluation.type,
    },
  }
}

export const jsExpression = ({
  name,
  label,
  help,
  multiline = true,
  canInterpolate,
  interpolate,
  evaluations,
  value,
}: {
  name: string,
  label?: string,
  help?: string,
  multiline?: boolean,
  canInterpolate?: boolean,
  interpolate?: boolean,
  evaluations?: Evaluation[],
  value?: string,
}): Param => {
  return {
    name,
    type: 'StringableParam',
    label: label ?? name,
    help: help ?? '',
    multiline: multiline ?? false,
    canInterpolate: canInterpolate ?? true,
    interpolate: interpolate ?? true,
    evaluations: evaluations ?? [
      jsonEvaluation,
      jsFunctionEvaluation,
      jsExpressionEvaluation,
    ],
    casts: [
      numberCast,
      stringCast,
    ],
    input: {
      rawValue: value ?? 0,
      Evaluation: jsExpressionEvaluation.type,
    },
  }
}
