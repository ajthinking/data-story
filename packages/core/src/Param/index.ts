import { PortName } from '../types/Port'
import { snakeCaseToTitleCase } from '../utils/snakeCaseToTitleCase'
import { Cast } from './Cast'
import { Evaluation } from './Evaluation'
import { numberCast } from './casts/numberCast'
import { stringCast } from './casts/stringCast'
import { hjsonEvaluation } from './evaluations/hjsonEvaluation'
import { jsEvaluation } from './evaluations/jsEvaluation'
import { jsonEvaluation } from './evaluations/jsonEvaluation'

/**
 * The mode types ****************************************
 */
export type Stringable = {
  type: 'Stringable',
  multiline: boolean,
  canInterpolate: boolean,
  interpolate?: boolean,
  interpolationsFromPort?: PortName[],
  casts?: Cast[],
  evaluations?: Evaluation[],
  value: string | number
}

export type PropertySelection = {
  type: 'PropertySelection',
  value: string
}

export type PortSelection = {
  type: 'PortSelection',
  allowCreate: boolean,
  value: string
}

export type Select = {
  type: 'Select',
  options: {
    label: string
    value: any
  }[],
  value: string
}

export type Repeatable<RepeatableRow> = {
  type: 'Repeatable',
  row: RepeatableRow,
  value: Param[]
}

export type InputMode =
  Select |
  Stringable |
  PropertySelection |
  PortSelection |
  Repeatable<Param[]>

/**
 * The param type ****************************************
 */
export type Param = {
  name: string,
  label: string,
  help: string,
  inputMode: InputMode,
}

export type ParamValue = Param['inputMode']['value']

// quick param builders

// Provides a param with sensible defaults
export const param = ({
  name,
  label,
  help,
  inputMode,
}: {
  name: string,
  label?: string,
  help?: string,
  inputMode?: {
    type?: 'Stringable',
    multiline?: boolean,
    canInterpolate?: boolean,
    interpolate?: boolean,
    casts?: Cast[],
    evaluations?: Evaluation[],
    value?: string,
  }
}): Param => {
  return {
    name,
    label: label || snakeCaseToTitleCase(name),
    help: help || '',
    inputMode: {
      type: 'Stringable',
      multiline: inputMode?.multiline || false,
      canInterpolate: inputMode?.canInterpolate || true,
      interpolate: inputMode?.interpolate || true,
      casts: inputMode?.casts || [], 
      value: inputMode?.value || '',
      evaluations: inputMode?.evaluations || [],
    },
  }
}

export const str = ({
  name,
  label,
  help,
  multiline,
  canInterpolate,
  interpolate,
  evaluations,
  value,
}: {
  name: string,
  label?: string,
  help?: string,
  multiline?: boolean,
  canInterpolate?: boolean,
  interpolate?: boolean,
  evaluations?: Evaluation[],
  value?: string,
}): Param => {
  return param({
    name,
    label,
    inputMode: {
      multiline: multiline,
      canInterpolate: canInterpolate ?? true,
      interpolate: interpolate ?? true,
      evaluations: evaluations,
      casts: [
        { ...stringCast, selected: true },
      ],
      value: value,
    }
  })
}

export const num = ({
  name,
  label,
  help,
  multiline,
  canInterpolate,
  interpolate,
  evaluations,
  value,
}: {
  name: string,
  label?: string,
  help?: string,
  multiline?: boolean,
  canInterpolate?: boolean,
  interpolate?: boolean,
  evaluations?: Evaluation[],
  value?: string,
}): Param => {
  return param({
    name,
    label,
    inputMode: {
      multiline: multiline,
      canInterpolate: canInterpolate ?? true,
      interpolate: interpolate ?? true,
      evaluations: evaluations,
      casts: [
        { ...numberCast, selected: true },
      ],
      value: value,
    }
  })
}

export const json_ = ({
  name,
  label,
  help,
  multiline,
  canInterpolate,
  interpolate,
  evaluations,
  value,
}: {
  name: string,
  label?: string,
  help?: string,
  multiline?: boolean,
  canInterpolate?: boolean,
  interpolate?: boolean,
  evaluations?: Evaluation[],
  value?: string,
}): Param => {
  return param({
    name,
    help,
    label,
    inputMode: {
      multiline: multiline ?? true,
      canInterpolate: canInterpolate ?? true,
      interpolate: interpolate ?? true,
      casts: [],
      evaluations: evaluations || [
        { ...jsonEvaluation, selected: true}
      ],
      value: value,
    }
  })
}