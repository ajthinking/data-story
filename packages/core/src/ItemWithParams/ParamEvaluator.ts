import { Param } from '../Param';
import { ItemValue } from '../types/ItemValue';
import { StringableParamEvaluator } from './StringableParamEvaluator';
import { ParamsValueEvaluator } from '../types/ParamsValueEvaluator';
import { RepeatableParamEvaluator } from './RepeatableParamEvaluator';
import { SelectableParamEvaluator } from './SelectableParamEvaluator';

export class ParamEvaluator implements ParamsValueEvaluator<any>{
  private evaluators: ParamsValueEvaluator<Param>[] = [
    new StringableParamEvaluator(),
    new SelectableParamEvaluator(),
    new RepeatableParamEvaluator(this),
  ]

  canEvaluate(param: Param): param is Param {
    return this.evaluators.some(e => e.canEvaluate(param));
  }

  evaluate(itemValue: ItemValue, param: Param, globalParams: Param[]) {
    const evaluator = this.evaluators.find(e => e.canEvaluate(param));
    if(!evaluator) return param.value;

    return evaluator.evaluate(itemValue, param, globalParams);
  }
}
