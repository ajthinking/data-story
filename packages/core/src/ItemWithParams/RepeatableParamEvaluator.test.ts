import { Param, RepeatableParam } from '../Param';
import { beforeEach, describe, expect, SpyInstance } from 'vitest';
import { ParamEvaluator } from './ParamEvaluator';
import { RepeatableParamEvaluator } from './RepeatableParamEvaluator';
import { mockPortMapData, mockRepeatableData, removePropertyData } from './mock';


describe('RepeatableParamEvaluator', () => {
  let spy: SpyInstance;
  let evaluator: RepeatableParamEvaluator;

  beforeEach(() => {
    const paramEvaluator = new ParamEvaluator();
    spy = vi.spyOn(paramEvaluator, 'evaluate');
    evaluator = new RepeatableParamEvaluator(paramEvaluator);
  });

  it('should be able to evaluate RepeatableParam', () => {

    expect(evaluator instanceof RepeatableParamEvaluator).toBe(true);
    expect(evaluator.canEvaluate({ type: 'RepeatableParam' } as unknown as Param)).toBe(true);
  });

  it('should test evaluate method performs when param.row length is 0', () => {
    const result = evaluator.evaluate({}, {
      'name': 'remove_properties',
      'type': 'RepeatableParam',
      'row': [],
      'value': []
    } as unknown as RepeatableParam<any>);

    expect(result).toEqual([]);
  });

  it('should test RepeatableParam contain the StringableParam type ', () => {
    const result = evaluator.evaluate({}, removePropertyData as unknown as RepeatableParam<any>);

    expect(result).toEqual(removePropertyData.value);
  });

  it('should test RepeatableParam contain the PortSelectionParam type', () => {
    const result = evaluator.evaluate({}, mockPortMapData as unknown as RepeatableParam<any>);

    expect(result).toEqual(mockPortMapData.value);

  });

  it('should test RepeatableParam contain the RepeatableParam type', () => {
    const result = evaluator.evaluate({}, mockRepeatableData as unknown as RepeatableParam<any>);

    expect(result).toEqual(mockRepeatableData.value);
  });

  describe('call evaluator recursively', () => {
    beforeEach(() => {
      evaluator.evaluate({}, mockRepeatableData as unknown as RepeatableParam<any>);
    });

    it('recursive called count', () => {

      /**
       * count = 6 = (1                +  1                +  1                             ) * 2
       *             ([StringableParam] + [RepeatableParam] + [inner nested StringableParam]) * 2[repeatable value]
       */
      expect(spy).toHaveBeenCalledTimes(6);
    });

    it('recursive called with correct params', () => {
      const calls = spy.mock.calls;
      expect(calls).toMatchSnapshot();
    });
  })
});
