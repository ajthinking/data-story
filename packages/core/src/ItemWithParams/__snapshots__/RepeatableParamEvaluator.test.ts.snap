// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RepeatableParamEvaluator > call evaluator recursively > recursive called with correct params 1`] = `
[
  [
    {},
    {
      "name": "value",
      "type": "StringableParam",
      "value": "value-11",
    },
  ],
  [
    {},
    {
      "name": "remove_properties",
      "row": [
        {
          "name": "property",
          "type": "StringableParam",
          "value": "id",
        },
      ],
      "type": "RepeatableParam",
      "value": [
        {
          "property": "property-11",
        },
      ],
    },
  ],
  [
    {},
    {
      "name": "property",
      "type": "StringableParam",
      "value": "property-11",
    },
  ],
  [
    {},
    {
      "name": "value",
      "type": "StringableParam",
      "value": "value-22",
    },
  ],
  [
    {},
    {
      "name": "remove_properties",
      "row": [
        {
          "name": "property",
          "type": "StringableParam",
          "value": "id",
        },
      ],
      "type": "RepeatableParam",
      "value": [
        {
          "property": "property-22",
        },
      ],
    },
  ],
  [
    {},
    {
      "name": "property",
      "type": "StringableParam",
      "value": "property-22",
    },
  ],
]
`;
